ref/rust/src/segwit_addr.rs

```diff
@@ -1,7 +1,174 @@
-pub fn encode() {
-    
+use bech32;
+
+/// Witness `WitnessProgram` type
+#[derive(Debug)]
+pub struct WitnessProgram {
+    /// Witness program version
+    pub version: u8,
+    /// Witness program content
+    pub program: Vec<u8>
+}
+
+impl WitnessProgram {
+    /// Converts a `WitnessProgram` to a script public key
+    ///
+    /// The format for the output is 
+    /// `[version, program length, <program>]`
+    pub fn to_scriptpubkey(&self) -> Vec<u8> {
+        let mut pubkey: Vec<u8> = Vec::new();
+        let mut v = self.version;
+        if v > 0 {
+            v += 0x80;
+        }
+        pubkey.push(v);
+        pubkey.push(self.program.len() as u8);
+        pubkey.extend_from_slice(&self.program);
+        pubkey
+    }
+
+    pub fn from_scriptpubkey(pubkey: Vec<u8>) -> Result<WitnessProgram, ()> {
+        // TODO: Validate according to BIP141
+        
+        // We need a version byte and a program length byte, with a program at 
+        // least 2 bytes long.
+        if pubkey.len() < 4 {
+            return Err(())
+        }
+        let proglen: usize = pubkey[1] as usize;
+        // Check that program length byte is consistent with pubkey length
+        if pubkey.len() != 2 + proglen {
+            return Err(())
+        }
+        // Process script version
+        let mut v: u8 = pubkey[0];
+        if v > 0x80 {
+            v -= 0x80;
+        }
+        let (_, program) = pubkey.split_at(2);
+        Ok(WitnessProgram {
+            version: v,
+            program: program.to_vec()
+        })
+    }
+
+    pub fn clone(&self) -> WitnessProgram {
+        WitnessProgram {
+            version: self.version,
+            program: self.program.to_vec()
+        }
+    }
 }
 
-pub fn decode() {
-    
+pub fn encode(hrp: String, p: WitnessProgram) -> EncodeResult {
+    let mut data: Vec<u8> = vec![p.version];
+    // Convert 8-bit program into 5-bit
+    let p5 = match convert_bits(p.program.to_vec(), 8, 5, true) {
```

Comment:
Since `convert_bits` is used for both encoding and decoding, it returns a result type containing a `CodingError`.

Elsewhere you mentioned a single `Error` type. I may have gone overboard in defining lots of errors, but the error hierarchy may be useful. Each function may return each `enum` value for its error type and no more or less. The alternative would be a flat `Error` struct with all error values in it. This would enable code shortcuts like `try!` and `?`. Would you recommend restructuring errors to be a flat `enum`?